omnisolver/samplers/van.py

```diff
@@ -0,0 +1,188 @@
+                    norm_type = 2
+                    total_norm = 0
+                    for p in parameters:
+                        param_norm = p.grad.data.norm(norm_type)
+                        total_norm += param_norm.item()**norm_type
+                        total_norm = total_norm**(1 / norm_type)
+                        clip_coef = max_norm / (total_norm + self.epsilon)
+                        for p in parameters:
+                            p.grad.data.mul_(clip_coef)
+
+                optimizer.step()
+
+                train_time += time.time() - train_start_time
+
+                if self.print_step and step % self.print_step == 0:
+                    free_energy_mean = loss.mean() / beta / self.n
+                    free_energy_std = loss.std() / beta / self.n
+                    entropy_mean = -log_prob.mean() / self.n
+                    energy_mean = energy.mean() / self.n
+                    mag = sample.mean(dim=0)
+                    mag_mean = mag.mean()
+                    if step > 0:
+                        sample_time /= self.print_step
+                        train_time /= self.print_step
+                    used_time = time.time() - start_time
+                    self.my_log(
+                        'beta = {:.3g}, # {}, F = {:.8g}, F_std = {:.8g}, S = {:.5g}, E = {:.5g}, M = {:.5g}, sample_time = {:.3f}, train_time = {:.3f}, used_time = {:.3f}'
+                        .format(
+                            beta,
+                            step,
+                            free_energy_mean.item(),
+                            free_energy_std.item(),
+                            entropy_mean.item(),
+                            energy_mean.item(),
+                            mag_mean.item(),
+                            sample_time,
+                            train_time,
+                            used_time,
+                        ))
+                    sample_time = 0
+                    train_time = 0
+
+            with open(self.fname, 'a', newline='\n') as f:
+                f.write('{} {} {:.3g} {:.8g} {:.8g} {:.8g} {:.8g}\n'.format(
+                    self.n,
+                    self.seed,
+                    beta,
+                    free_energy_mean.item(),
+                    free_energy_std.item(),
+                    energy_mean.item(),
+                    entropy_mean.item(),
+                ))
+
+            beta += self.beta_inc
+
+        return dimod.SampleSet.from_samples(
+            sample, energy=energy, vartype=bqm.vartype
+        )
+
+    @property
+    def properties(self):
+        return dict()
+
+    @property
+    def parameters(self):
+        return {"num_reads": []}
+
+
+class ModifiedSKModel(SKModel):
+    def __init__(self, n, device, bqm, field=0, seed=0):
+        self.n = n
+        self.field = field
+        self.seed = seed
+        if seed > 0:
+            torch.manual_seed(seed)
+        J = bqm.to_numpy_matrix()
```

Comment:
`to_numpy_matrix()` does not return couplings matrix, it converts BQM to Qubo and returns a matrix of the associated quadratic form. Considering how the energy is computed in the original `SKModel`, are you sure that `ModifiedSKModel` computes energy correctly?